File: android/app/src/main/java/app/systemresponse/PacketQueue.kt
package app.systemresponse

/** Coalesce unsent snapshots; encode only at dequeue to preserve authenticated ordering. */
class PacketQueue {
    private val commands = ArrayDeque<Packet>()
    private val snapshots = linkedMapOf<String, Packet>()
    val size get() = commands.size + snapshots.size
    fun add(packet: Packet): Boolean {
        if (packet.type in setOf("activity", "autoStatus", "ping", "pong")) snapshots[packet.type] = packet
        else { if (commands.size >= 32) return false; commands.addLast(packet) }
        return true
    }
    fun next(): Packet? {
        if (commands.isNotEmpty()) return commands.removeFirst()
        val key = snapshots.keys.firstOrNull() ?: return null
        return snapshots.remove(key)
    }
    fun clear() { commands.clear(); snapshots.clear() }
}

File: macos/Sources/PacketQueue.swift
import Foundation

/// Coalesce only unsent snapshots. Authenticate at dequeue so sequence numbers stay ordered.
struct PacketQueue {
    private var commands: [Packet] = []
    private var snapshots: [Packet] = []
    var count: Int { commands.count + snapshots.count }
    mutating func append(_ packet: Packet) -> Bool {
        if ["activity", "autoStatus", "ping", "pong"].contains(packet.type) {
            snapshots.removeAll { $0.type == packet.type }
            snapshots.append(packet)
        } else {
            guard commands.count < 32 else { return false }
            commands.append(packet)
        }
        return true
    }
    mutating func next() -> Packet? {
        if !commands.isEmpty { return commands.removeFirst() }
        return snapshots.isEmpty ? nil : snapshots.removeFirst()
    }
    mutating func reset() { commands.removeAll(); snapshots.removeAll() }
}
